### DISCUSSION TITLE
Feedback over "Argon Pro" Combo Colors

### DISCUSSION THREAD

**Discussion by HeyZeer0** [ppy/osu#21905] [Ideas]
After playing a few maps using Argon Pro, I noticed a very clear difficulty in sight-reading maps.
Digging a little more I'm pretty confident the issue is related to the `Color Contrast` between the background and the colors used.

## What is color contrast
Color contrast is the contrast ratio between two colors and can be calculated by dividing the luminance of both colors. The lower the contrast ratio, the more effort the human brain needs to do to differentiate both colors making reading texts or in this case circles harder. Unfortunately, I'm not a specialist on the topic but there is [this awesome thread by Justin Baker](https://medium.muz.li/the-science-of-color-contrast-an-expert-designers-guide-33e84c41d156) that explains all the calculations involved and some of the human behavior behind it.

For further clarification here is a low-contrast combination versus a good one, in order.
You will notice that both are readable but you can clearly notice that the high-contrast one is easier to read

#### Low Contrast (3.09 score)
![Low Contrast](https://cdn.hey0.dev/ss/2022/12/28165024.png)

#### High Contrast (17.36 score)
![High Contrast](https://cdn.hey0.dev/ss/2022/12/28165228.png)

## How does this affect Argon?
Currently, on Argon there are 6 different combo colors and 4 of them have a bad contrast ratio. Alongside that, the contrast between the background and the approach circle is considerably worse, making it harder to read them. The problem is really clear with the blue and red colors, which are by far the worse ones (in my opinion). Of course, this is taking into consideration a completely black background since that is how most "pro" players play and using **approximated values for the colors**.

Here is a list of all combo colors and their contrast ratio:

#### Green | Circle Score: `12.59` | Approach Circle Score: `7.17`
![Green hit circle](https://cdn.hey0.dev/ss/2022/12/28170120.png)

#### Blue | Circle Score: `3.45` | Approach Circle Score: `2.18`
![Blue hit circle](https://cdn.hey0.dev/ss/2022/12/28170214.png)

#### Purple | Circle Score: `2.77` | Approach Circle Score: `1.89`
![Purple hit circle](https://cdn.hey0.dev/ss/2022/12/28170344.png)

#### Yellow | Circle Score: `14.78` | Approach Circle Score: `8.04`
![Yellow hit circle](https://cdn.hey0.dev/ss/2022/12/28170506.png)

#### Orange | Circle Score: `6.79` | Approach Circle Score: `5.04`
![Orange hit circle](https://cdn.hey0.dev/ss/2022/12/28170604.png)

#### Red | Circle Score: `4.85` | Approach Circle Score: `4.32`
![Red hit circle](https://cdn.hey0.dev/ss/2022/12/28170826.png)

## Finalizing
The solution for this issue is simply improving the color contrast of all combo colors, which will make reading maps severely easier.
In any case massive shoutout to the Osu! team for putting effort into creating this awesome skin! You guys are on the right path and I look forward to seeing it even better!

**Comment by mcpower** [comment]
You can adjust the combo colours to be more bright in all skins using the "combo color normalisation" option in Visual Options, implemented in #20736. Check out the video [lazer updates - November 17, 2022](https://youtu.be/RYLD1CJSBvU?t=1m32s) for an explanation.

**Reply by HeyZeer0** [reply]
Interesting feature although it does not solve the problem. In my opinion, it does not make sense for a default skin to be only "playable" when a specific option is enabled, especially when this is not about dark combo colors. This is not a bug report, it is a suggestion for a welcome design change 😄

**Comment by mk56-spn** [comment]
i think this is unnecessarily dumbing down the skin for a miniscule gain, there is and has always been plenty of pro players out there that have skins with dull or not contrast maxxed default combo colours (whitecat, rohulk, khz...)  if you want to count ones with similar colours to argon or that have the same outer white ring with inner colour design then theres also bubbleman's skins , seoul v10 skin.. i could go on but my point is i think this enters the " optimising the fun out of the game " territory even for pro skins, i mean anyone who really wants the most optimal skin out there for situations where immediate visibility is critical, like high AR, will probably already be opting for something like  the rafis skin ,or one of its many variants

**Reply by HeyZeer0** [reply]
Thanks for the feedback! 
This is not a case of "optimizing the fun out of the game" its about accessibility end how the human brain reads colors. Of course, you can have all the colors as is and they are sorta readable, but that doesn't mean they are at their best and neither that you need dull colors for it to work: this is not a contrast-only issue, it is a contrast ratio against the dark background issue. As I showed earlier you can have high-contrast colors that have good readability which is the case for the green combo color. Argon has the potential
to be an awesome default skin and I believe increasing the readability of the circles will always be a good thing ^^

Just as an extra, the gain is not minuscule, it is quite considerable when you are reading faster-paced maps or maps that have a lot of hit circles, being able to clearly read the circles is a core game mechanic.

**Comment by MohaMG25** [comment]
I’ve always had this feeling that something is wrong with the argon skin. The colours are cool and pretty but its somehow effecting the gameplay negatively and maybe the “colour contrast” is the problem

**Comment by yamplum** [comment]
I agree that contrast is a pretty big problem. Blue and purple especially are subjectively very hard to read (purple might as well be bargain bin Hidden) and I'm glad to see the numbers reflect that. I get that argon is going for a vibrant aesthetic but the colors end up overly saturated.

For readability I think I would prefer lighter colors, something like [this map's](https://osu.ppy.sh/beatmapsets/374900#osu/827835) (below). Though customizable colors would be ideal, of course.

https://user-images.githubusercontent.com/118788705/210287312-1ffb40e8-79e9-460b-b3ac-3b2c8e6ddfdd.mp4

**Reply by HeyZeer0** [reply]
Can agree, that is severely better and still beautiful!

**Comment by peppy** [comment]
Argon was not made to be played against a black background.

Argon "Pro" only changes hitsounds and removes "Great"s for now. We may change the combo colours in the future, or just wait for user-level customisation via the skin editor.

Also I have no idea why you're testing the approach circle It's the same colour as the hitcircle unless you are measuring at different points in the fade process (which you are, judging from screenshots). But that's unimportant.

**Reply by yamplum** [reply]
> Argon was not made to be played against a black background.

Unfortunately, playing with a non-black background doesn't help much.

I see that argon uses 100% saturation colors for its combos. There is a pretty steep gradient in luminosity between fully saturated colors of different hues, with reds and blues being the darkest (corresponding to the lack of receptor cells covering those parts of the spectrum).

![image](https://user-images.githubusercontent.com/118788705/210305054-7a3af2d1-8e9c-4f2b-b5c0-f94070403c46.png)

You can see that the darker colors are perceived as medium gray in luminosity. This means that both very bright and very dark backgrounds would be the best case conditions for maximum contrast (extreme values being most distant from the middle). Default dim settings, however, create a background that's closer to medium gray, resulting in significantly reduced contrast.

For example, purple combo color from best to worst contrast (at least subjectively):
<details>
<summary>Click to expand</summary>

![osu!_DOkg3NEy0A](https://user-images.githubusercontent.com/118788705/210305850-1c82341d-8b5f-438d-beda-00cf7bd106aa.png)
![osu!_EBryo3z8Lu](https://user-images.githubusercontent.com/118788705/210305854-c0b81518-30ea-45ec-83e3-edfa04f501e0.png)
![osu!_8k5SLPtT0g](https://user-images.githubusercontent.com/118788705/210305860-3cc56688-4429-44e5-be84-9026e2b78d4b.png)
![osu!_bVpitKRxIe](https://user-images.githubusercontent.com/118788705/210305863-36fd8b3b-6ddb-4f27-9f17-9761a477fbf2.png)
</details>

I think the correct approach here would be to choose colors of similar luminosity rather than saturation to compensate the differences in perception, and make them reasonably distant from the target background luminosity.

One extra thing that's hard to convey in screenshots is that approach circles are quickly moving, which produces heavy motion blur that further affects their visibility (and they're translucent as well). The choice of combo colors is crucial to the gameplay performance of the skin, so I hope this will be taken into account. Of course, this affects the default non-pro skin as well.

**Reply by HeyZeer0** [reply]
> Argon was not made to be played against a black background.

I see, but yeah can agree with @yamplum, playing with a background is quite hard, especially because of the difference in color contrasts between the combos and the actual background. Players play with a black background because it makes the circles easier to see.

> Also I have no idea why you're testing the approach circle It's the same colour as the hitcircle unless you are measuring at different points in the fade process (which you are, judging from screenshots). But that's unimportant.

Yes! the approach circle has the same color but they have a different alpha based on how close they are to the hit circle making them blend with the background. Taking that into consideration you can say that for most of the "travel time" of the approach circle, you will be seeing it dimmed reducing the overall visibility. Although I can agree that in most cases it might not be as relevant as the circles themselves.

Allowing further skin customization of the combo colors would solve the issue in all cases, but would also be nice to have friendlier colors as default option 😄
